python_modules/dagster/dagster/_core/definitions/events.py

```diff
@@ -548,100 +547,6 @@ def metadata(self) -> MetadataMapping:
         return {entry.label: entry.value for entry in self.metadata_entries}
 
 
-@whitelist_for_serdes
-class Materialization(
-    NamedTuple(
-        "_Materialization",
-        [
-            ("label", str),
-            ("description", Optional[str]),
-            ("metadata_entries", Sequence[MetadataEntry]),
-            ("asset_key", AssetKey),
```

Comment:
Context: 
- `Materialization` which did not require asset key was deprecated in `0.10.0` (https://dagster.phacility.com/D3866)
- `asset_key` was switched from optional to non-optional in `0.12.10` (https://dagster.phacility.com/D9139)

This might be a moot point since it's been this way for so long, but I know that the `Materialization` event preceded the concept of `AssetKey`, so I'm wondering if this will change behavior for deserialization of materialization events persisted without asset keys (pre-`0.10.0`).  

The way I'm reading it is that we've already broken everyone who has stored legacy Materialization events, in which case I'm fine with this change, but could use another brain to think through this, since I don't want our backcompat handling to change with this PR.

Reply:
>The way I'm reading it is that we've already broken everyone who has stored legacy Materialization events, in which case I'm fine with this change, but could use another brain to think through this, since I don't want our backcompat handling to change with this PR.

If `AssetKey` became non-optional on `Materialization` in the past, then yes, it would've already broken anyone storing a `Materialization` without an `AssetKey`, given that there is no custom serializer logic for `Materialization`. So I don't think this PR changes anything there... requesting @alangenfeld for second take.

Reply:
while the static typing was off  - it appears to me looking at `[1]` that we would deserialize a `Materialization` with `asset_key` `None` at runtime just fine


through it turns out `AssetMaterialization` does not runtime check against `None` currently either (looks like an accident from https://github.com/dagster-io/dagster/pull/10496) 
```
>>> AssetMaterialization(asset_key=None)
AssetMaterialization(asset_key=None, description=None, metadata_entries=[], partition=None, tags=None)
```

Not sure exactly where that leaves us. I think if we don't fix `AssetMaterialization` then its fine - but if we do then its not? Not sure we can fix `AssetMaterialization` without it being a breaking change unless something else downstream fails on `None` `asset_key`

Reply:
> Since it was operational before then, any old Materializations with null asset keys would've been breaking a few months ago. So we're not going to be newly breaking old materialization events if we reintroduce that check now.

I don't follow this - do we coerce `Materialization` in to `AssetMaterialization` somewhere?

Reply:
My mistake, you are right. Re-introducing null-check on `AssetMaterialization` + this PR (which introduces the coercion that was absent before) would break old asset-key-less `Materialization` objects that deserialized fine before.

But IIUC:

- This PR as is does not introduce behavior change because `AssetMaterialization` in its current state does _not_ null-check `AssetKey`
- A reasonable solution to the problem above is to introduce a default dummy asset key

Reply:
got it - lets put it under test [2]